### PR TITLE
[CoreText] Remove the Foundation dependency from CTFramesetter and CTTypeSetter with code improvements

### LINKED ISSUE #2355: CTTypesetter should be implemented via CFRuntimeBase (and not an object)

### PR DESCRIPTION
**Pull request by msft-Jeyaram**
- Implement CTFramesetter and CTTypeSetter via Runtimebase
- Remove member access of CTFramesetter & CTTypeSetter
- Code improvements

Fixes #2355 and  #2358

### CODE REVIEW

File: Frameworks/include/CoreTextInternal.h [hunk 1]
```diff
@@ -17,7 +17,7 @@
 
 #import <CoreText/CoreText.h>
 #import <CoreText/CTParagraphStyle.h>
-
+#import <CFCppBase.h>
```
Comment (msft-Jeyaram) [inline Frameworks/include/CoreTextInternal.h]: this can go away

File: Frameworks/CoreText/CTFramesetter.mm [hunk 2]
```diff
@@ -19,36 +19,45 @@
...
+
     CGRect frameRect = CGPathGetBoundingBox(path);
-    _CTFramesetter* framesetter = static_cast<_CTFramesetter*>(framesetterRef);
 
     // Call _DWriteWrapper to get _CTLine object list that makes up this frame
-    _CTTypesetter* typesetter = static_cast<_CTTypesetter*>(framesetter->_typesetter);
+    CFAttributedStringRef attributedString = __CTTypesetterGetAttributedString(framesetter->Typesetter());
     if (range.length == 0L) {
-        range.length = [typesetter->_string length] - range.location;
+        range.length = CFStringGetLength(CFAttributedStringGetString(attributedString)) - range.location;
```
Comment (DHowett-MSFT) [inline Frameworks/CoreText/CTFramesetter.mm]: why not CFAttributedStringGetLength? Technically, an NSAttributedString subclass can override `-length`, and with the way you've implemented this, it will not get called.

File: Frameworks/CoreText/CTFramesetter.mm [hunk 3]
```diff
@@ -19,36 +19,45 @@
...
-    RETURN_NULL_IF(framesetterRef == nil || path == nullptr);
+CTFrameRef CTFramesetterCreateFrame(CTFramesetterRef framesetter, CFRange range, CGPathRef path, CFDictionaryRef frameAttributes) {
+    RETURN_NULL_IF(framesetter == nil || path == nullptr);
+
     CGRect frameRect = CGPathGetBoundingBox(path);
-    _CTFramesetter* framesetter = static_cast<_CTFramesetter*>(framesetterRef);
 
     // Call _DWriteWrapper to get _CTLine object list that makes up this frame
-    _CTTypesetter* typesetter = static_cast<_CTTypesetter*>(framesetter->_typesetter);
+    CFAttributedStringRef attributedString = __CTTypesetterGetAttributedString(framesetter->Typesetter());
```
Comment (DHowett-MSFT) [inline Frameworks/CoreText/CTFramesetter.mm]: Double underscores are for file-private static functions in the CT convention.

File: Frameworks/CoreText/CTTypesetter.mm [hunk 4]
```diff
@@ -92,22 +90,25 @@ CTLineRef CTTypesetterCreateLineWithOffset(CTTypesetterRef ts, CFRange range, do
...
 */
 CFIndex CTTypesetterSuggestLineBreakWithOffset(CTTypesetterRef ts, CFIndex index, double width, double offset) {
-    _CTTypesetter* typesetter = static_cast<_CTTypesetter*>(ts);
-    _CTFrame* frame = _DWriteGetFrame(static_cast<CFAttributedStringRef>(typesetter->_attributedString.get()),
-                                      CFRangeMake(index, [typesetter->_string length] - index),
-                                      CGRectMake(offset, 0, width, FLT_MAX));
-    return ([frame->_lines count] > 0) ? static_cast<_CTLine*>([frame->_lines firstObject])->_strRange.length : 0;
+    __CTTypesetter* typesetter = const_cast<__CTTypesetter*>(ts);
+    _CTFrame* frame =
+        _DWriteGetFrame(__CTTypesetterGetAttributedString(typesetter),
```
Comment (DHowett-MSFT) [inline Frameworks/CoreText/CTTypesetter.mm]: this one can use the internal function `->AttributedString()`

File: Frameworks/CoreText/CTTypesetter.mm [hunk 5]
```diff
@@ -92,22 +90,25 @@ CTLineRef CTTypesetterCreateLineWithOffset(CTTypesetterRef ts, CFRange range, do
...
+                        CFRangeMake(index, CFStringGetLength(CFAttributedStringGetString(typesetter->AttributedString())) - index),
```
Comment (DHowett-MSFT) [inline Frameworks/CoreText/CTTypesetter.mm]: same as in CTFramesetter

File: Frameworks/CoreText/CTFramesetter.mm (hunk 2 above)
Comment (msft-Jeyaram) [inline Frameworks/CoreText/CTFramesetter.mm]: The initial implementation was pretty much this. get the string out of the attributedString then get the length.
you are correct, needs to be updated.

---
In reply to: [110286467](https://github.com/Microsoft/WinObjC/pull/2437#discussion_r110286467) [](ancestors = 110286467)

File: Frameworks/CoreText/CTTypesetter.mm [hunk 6]
```diff
@@ -92,22 +90,23 @@ CTLineRef CTTypesetterCreateLineWithOffset(CTTypesetterRef ts, CFRange range, do
...
-CFIndex CTTypesetterSuggestLineBreakWithOffset(CTTypesetterRef ts, CFIndex index, double width, double offset) {
-    _CTTypesetter* typesetter = static_cast<_CTTypesetter*>(ts);
-    _CTFrame* frame = _DWriteGetFrame(static_cast<CFAttributedStringRef>(typesetter->_attributedString.get()),
-                                      CFRangeMake(index, [typesetter->_string length] - index),
+CFIndex CTTypesetterSuggestLineBreakWithOffset(CTTypesetterRef typesetter, CFIndex index, double width, double offset) {
+    _CTFrame* frame = _DWriteGetFrame(typesetter->AttributedString(),
+                                      CFRangeMake(index, CFAttributedStringGetLength(typesetter->AttributedString()) - index),
                                       CGRectMake(offset, 0, width, FLT_MAX));
-    return ([frame->_lines count] > 0) ? static_cast<_CTLine*>([frame->_lines firstObject])->_strRange.length : 0;
+    if ([frame->_lines count] > 0) {
```
Comment (rajsesh-msft) [inline Frameworks/CoreText/CTTypesetter.mm]: >[frame->_lines count] [](start = 8, length = 21)

These objc calls can already be changed to CFCalls due to bridging so you dont have to touch this again when you fix CTLines

Comment (msft-Jeyaram) [inline Frameworks/CoreText/CTTypesetter.mm]: I've already done them in a separate branch, i'll  merge that in when it's done.
Main reason it's not part of this cr is that i would have to add extra castings, which would get removed in the final sweep. so waiting till then.

---
In reply to: [110446875](https://github.com/Microsoft/WinObjC/pull/2437#discussion_r110446875) [](ancestors = 110446875)

**Comment by DHowett-MSFT** [comment]
nit: can you reword this to be <72 characters
